### PR TITLE
Fix/debug stream ovlerload for alignments and tuple

### PR DESCRIPTION
**Pull request by rrahn**
fixes seqan/product_backlog#125

I did not added complete documentation as I would first need some feedback from @marehr if there is more to the problem that needs to be fixed in a different way.

### CODE REVIEW

File: include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp [hunk 1]
```diff
@@ -472,8 +472,8 @@ std::ranges::iterator_t<range_type> erase_gap(range_type & rng,
     return rng.erase_gap(first, last);
 }
 //!\}
-
-namespace detail
+} // namespace seqan3
+namespace seqan3::detail
```
Comment (rrahn) [inline include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:476]:
```suggestion
} // namespace seqan3

namespace seqan3::detail
```

File: include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp [hunk 2]
```diff
@@ -563,8 +563,12 @@ template <typename tuple_t, typename char_t>
 //!\endcond
 inline debug_stream_type<char_t> & operator<<(debug_stream_type<char_t> & stream, tuple_t && alignment)
 {
-    static_assert(std::tuple_size_v<remove_cvref_t<tuple_t>> >= 2, "An alignment requires at least two sequences.");
-    detail::stream_alignment(stream, alignment, std::make_index_sequence<std::tuple_size_v<remove_cvref_t<tuple_t>> - 1> {});
+    using alignment_t = remove_cvref_t<tuple_t>;
+    constexpr size_t number_of_sequences = std::tuple_size_v<alignment_t>;
```
Comment (rrahn) [inline include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:566]:
```suggestion
    constexpr size_t number_of_sequences = std::tuple_size_v<remove_cvref_t<tuple_t>>;
```

File: include/seqan3/core/concept/tuple.hpp [hunk 3]
```diff
@@ -186,3 +187,11 @@ SEQAN3_CONCEPT tuple_like = detail::tuple_size<std::remove_reference_t<t>> && re
...
 
 } // namespace seqan3
+
+namespace seqan3::detail
+{
+template <typename tuple_t>
+SEQAN3_CONCEPT debug_streamable_tuple = (!std::ranges::input_range<tuple_t>) &&
+                                        (!alphabet<tuple_t>)
+                                        && tuple_like<tuple_t>;
+} // namespace seqan3
```
Comment (rrahn) [inline include/seqan3/core/concept/tuple.hpp:197]:
```suggestion
} // namespace seqan3::detail
```

File: test/unit/core/detail/debug_stream_type_tuple_alignment_test.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include <sstream>
+#include <vector>
+
+#include <seqan3/alphabet/adaptation/char.hpp>
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/core/detail/debug_stream_tuple.hpp>
+
+// When including aligned_sequence_concept and debug_stream_tuple the overload for the stream operator
+// is ambiguous. The solution is to introduce a named concept for a debug streamble tuple type.
+TEST(debug_stream_type, incompatible_debbug_stream_overload)
```
Comment (rrahn) [inline test/unit/core/detail/debug_stream_type_tuple_alignment_test.cpp:19]:
```suggestion
// https://github.com/seqan/product_backlog/issues/125
TEST(debug_stream_type, issue_125)
```

File: include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp [hunk 5]
```diff
@@ -554,15 +557,16 @@ inline bool constexpr all_satisfy_aligned_seq<type_list<elems...>> = (aligned_se
...
-             (!alphabet<tuple_t>) && // exclude alphabet_tuple_base
-             tuple_like<remove_cvref_t<tuple_t>> &&
+    requires detail::debug_streamable_tuple<tuple_t> &&
              detail::all_satisfy_aligned_seq<detail::tuple_type_list_t<remove_cvref_t<tuple_t>>>
 //!\endcond
 inline debug_stream_type<char_t> & operator<<(debug_stream_type<char_t> & stream, tuple_t && alignment)
 {
-    static_assert(std::tuple_size_v<remove_cvref_t<tuple_t>> >= 2, "An alignment requires at least two sequences.");
-    detail::stream_alignment(stream, alignment, std::make_index_sequence<std::tuple_size_v<remove_cvref_t<tuple_t>> - 1> {});
+    constexpr size_t number_of_sequences = std::tuple_size_v<remove_cvref_t<tuple_t>>;
```
Comment (marehr) [inline include/seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp:565]:
```suggestion
    constexpr size_t sequences_size = std::tuple_size_v<remove_cvref_t<tuple_t>>;
```

File: include/seqan3/alignment/aligned_sequence/debug_stream_alignment.hpp [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * \param[in,out] stream The target stream for the formatted output.
+ * \param[in] alignment The alignment that shall be formatted. All sequences must be equally long.
+ *
+ * \return The given stream to which the alignment representation is appended.
+ */
+template <typename char_t, typename alignment_t>
+//!\cond
+    requires (detail::debug_streamable_tuple<alignment_t> &&
+              detail::all_satisfy_aligned_seq<detail::alignment_type_list_t<remove_cvref_t<alignment_t>>>)
```
Comment (marehr) [inline include/seqan3/alignment/aligned_sequence/debug_stream_alignment.hpp:101]: `alignment_type_list_t` is unknown, did you forget to add something?

File: include/seqan3/core/detail/debug_stream_tuple.hpp [hunk 7]
```diff
@@ -37,6 +39,22 @@ void print_tuple(debug_stream_type<char_t> & s, tuple_t && t, std::index_sequenc
...
+ * \details
+ *
+ * This concept requires that the given type is a seqan3::tuple_like type but neither an std::ranges::input_range nor
+ * an alphabet (see seqan3::alphabet_tuple_base).
+ */
+//!\cond
+template <typename tuple_t>
+SEQAN3_CONCEPT debug_streamable_tuple = (!std::ranges::input_range<tuple_t>) &&
+                                        (!alphabet<tuple_t>) &&  // exclude alphabet_tuple_base
+                                         tuple_like<tuple_t>;
```
Comment (marehr) [inline include/seqan3/core/detail/debug_stream_tuple.hpp:56]: In named concepts we don't need parenthesis.

Comment (marehr) [inline include/seqan3/core/detail/debug_stream_tuple.hpp:56]:
```suggestion
                                        tuple_like<tuple_t>;
```

Comment (marehr) [inline include/seqan3/core/detail/debug_stream_tuple.hpp:56]:
```suggestion
SEQAN3_CONCEPT debug_streamable_tuple = (!std::ranges::input_range<tuple_t>) &&
                                        (!alphabet<tuple_t>) &&  // exclude alphabet_tuple_base
                                        tuple_like<tuple_t>;
```

File: include/seqan3/alignment/aligned_sequence/debug_stream_alignment.hpp (hunk 6 above)
Comment (marehr) [reply include/seqan3/alignment/aligned_sequence/debug_stream_alignment.hpp:101]: Changing this to 

```suggestion
              detail::all_satisfy_aligned_seq<detail::tuple_type_list_t<remove_cvref_t<alignment_t>>>)
```

Did not compile for me on gcc9 and gcc10.

**Comment by rrahn** [comment]
@smehringer polite ping

**Pull request merged**